src/Page/Form.purs

```diff
@@ -47,18 +54,105 @@ component =
 
-  eval (EvalForm form a) = a <$ do
-    H.modify_ _ { form = pure form }
-    pure unit
+  eval (EvalForm expr a) = a <$ do
+    let evaled' = evalPage get expr
+        get key = Data.Map.lookup key values
+        values = keysPage expr
+    H.modify_ _ { values = values }
+    case evaled' of
+      Left _ -> pure unit
+      Right evaled -> H.modify_ _ { form = pure { expr, evaled } }
+
+  eval (UpdateKey key val a) = do
+    form <- H.gets _.form
+    case form of
+      Left _ -> pure a
+      Right { expr: expr' } -> do
+        H.modify_ \state ->
+          state { values = Data.Map.insert key val state.values }
+        let expr = setPage key val expr'
+        eval (EvalForm expr a)
+
+  evalPage :: (Key -> Maybe ExprType) -> Page Expr -> Either EvalError (Page ExprType)
+  evalPage get page = do
+    contents <- traverse (evalSection get) page.contents
+    pure page { contents = contents }
+
+  evalSection :: (Key -> Maybe ExprType) -> Section Expr -> Either EvalError (Section ExprType)
+  evalSection get section = do
+    contents <- traverse (evalField get) section.contents
+    pure section { contents = contents }
+
+  evalField :: (Key -> Maybe ExprType) -> Field Expr -> Either EvalError (Field ExprType)
+  evalField get field = do
+    description <- evalExpr get field.description
+    input <- evalInput get field.input
+    name <- evalExpr get field.name
+    visibility <- evalExpr get field.visibility
+    pure { description, key: field.key, input, name, visibility }
+
+  evalInput :: (Key -> Maybe ExprType) -> Input Expr -> Either EvalError (Input ExprType)
+  evalInput get = case _ of
+    Text input -> do
+      default <- traverse (evalExpr get) input.default
+      maxLength <- traverse (evalExpr get) input.maxLength
+      minLength <- traverse (evalExpr get) input.minLength
+      placeholder <- evalExpr get input.placeholder
+      required <- evalExpr get input.required
+      value <- evalInputState get input.value
+      pure
+        ( Text
+          { default
+          , maxLength
+          , minLength
+          , placeholder
+          , required
+          , value
+          }
+        )
+    Toggle input -> do
+      default <- traverse (evalExpr get) input.default
+      value <- evalInputState get input.value
+      pure (Toggle { default, value })
+
+  evalInputState :: (Key -> Maybe ExprType) -> Record (InputState Expr ()) -> Either EvalError (Record (InputState ExprType ()))
+  evalInputState get inputState = do
+    value <- traverse (evalExpr get) inputState.value
+    pure inputState { value = value }
+
+  keysPage :: Page Expr -> Map Key ExprType
+  keysPage page = foldMap keysSection page.contents
+
+  keysSection :: Section Expr -> Map Key ExprType
+  keysSection section = foldMap keysField section.contents
+
+  keysField :: Field Expr -> Map Key ExprType
+  keysField field = case field.input of
+    Text { value: { value: Just expr } }
+      | Right value <- evalExpr (const Nothing) expr ->
```

Comment:
❓ These `const Nothing`s are actually limiting the amount of depth we're able to have for lookups. If the expression we're looking up is a lookup itself, it won't find the key, even if it exists. Maybe there's a different way to do this? Maybe we should ditch the idea of a map in favor of the pub/sub stuff?

Reply:
Hmm, I'm not sure our lookups need to be recursive. I mean, that would be pretty fancy, but the results of our lookups will be the value of a form field, and I'm not really imagining allowing users to build up expressions in the generated form. Although...that would be pretty mind blowing 🤔

Reply:
Oh right, that makes sense! The thing we're looking up is an `ExprType` anyway, not an `Expr`. Sorry about that.